src/sql/src/query_model/hir/hir_from_qgm.rs

```diff
@@ -24,23 +31,393 @@ impl From<Model> for HirRelationExpr {
+    #[allow(dead_code)]
+    fn generate(mut self, mut model: Model) -> HirRelationExpr {
+        // Derive the RelationType of each box.
+        use crate::query_model::attribute::core::{Attribute, RequiredAttributes};
+        let attributes = HashSet::from_iter(std::iter::once(
+            Box::new(BoxRelationType) as Box<dyn Attribute>
+        ));
+        let root = model.top_box;
+        RequiredAttributes::from(attributes).derive(&mut model, root);
+
+        // Convert QGM to HIR.
         let _ = model.try_visit_pre_post(
-            &mut |model, box_id| -> Result<(), QGMError> {
-                println!("pre-visit");
+            &mut |_, _| -> Result<(), QGMError> {
+                // TODO: add to the context stack
                 Ok(())
             },
             &mut |model, box_id| -> Result<(), QGMError> {
-                println!("post-visit");
+                self.convert_subgraph(model, box_id);
                 Ok(())
             },
         );
-        // this should be todo!(), but we cannot merge to main code
-        // with todo!() statements
-        HirRelationExpr::constant(vec![vec![]], RelationType::empty())
+
+        // Retrieve the HIR and put `let`s around it.
+        let mut result = self.converted.pop().unwrap();
+        while let Some((id, value, _)) = self.lets.pop() {
+            if !matches!(value, HirRelationExpr::Get { .. }) {
+                result = HirRelationExpr::Let {
+                    name: "".to_string(),
+                    id,
+                    value: Box::new(value),
+                    body: Box::new(result),
+                }
+            }
+        }
+        result
+    }
+
+    /// Convert subgraph rooted at `box_id` to [HirRelationExpr].
+    ///
+    /// If the box corresponding to `box_id` has multiple ranging quantifiers,
+    /// the [HirRelationExpr] will be in `self.lets`, and it can be looked up in
+    /// `self.common_subgraphs`. Otherwise, it will be pushed to `self.converted`.
+    fn convert_subgraph(&mut self, model: &Model, box_id: &BoxId) {
+        let r#box = model.get_box(*box_id);
+        if self.common_subgraphs.get(&box_id).is_some() {
+            // Subgraph has already been converted.
+            return;
+        }
+        let hir = match &r#box.box_type {
+            BoxType::Get(Get { id, unique_keys }) => {
+                let typ = mz_repr::RelationType::new(
+                    r#box
+                        .columns
+                        .iter()
+                        .map(|c| {
+                            if let BoxScalarExpr::BaseColumn(BaseColumn { column_type, .. }) =
+                                &c.expr
+                            {
+                                column_type.clone()
+                            } else {
+                                // TODO: the validator should make this branch unreachable.
+                                panic!("expected all columns in Get BoxType to be BaseColumn");
```

Comment:
I think it's still better to return a user-facing error instead of `panic`. See the current state of `qgm_from_hir.rs` where some comments on top of `Err()` expressions in return positions indicate that those should be unreachable in practice. I suggest to reapeat that pattern here.